scripts/qa_crowbarsetup.sh

```diff
@@ -2304,30 +2326,39 @@ function oncontroller_testsetup()
-    nova volume-attach "$instanceid" "$volumeid" /dev/vdb | tee volume-attach.out
-    local volumeattachret=$?
-    device=`perl -ne "m!device [ |]*(/dev/\w+)! && print \\$1" volume-attach.out`
-    wait_for 9 5 "nova volume-show $volumeid | grep 'status.*in-use'" "volume to become attached" "volumeattachret=111"
-    ssh $vmip fdisk -l $device | grep 1073741824 || volumeattachret=$?
-    rand=$RANDOM
-    ssh $vmip "mkfs.ext3 -F $device && mount $device /mnt && echo $rand > /mnt/test.txt && umount /mnt"
-    nova volume-detach "$instanceid" "$volumeid" ; sleep 10
-    nova volume-attach "$instanceid" "$volumeid" /dev/vdb ; sleep 10
-    ssh $vmip fdisk -l $device | grep 1073741824 || volumeattachret=57
-    ssh $vmip "mount $device /mnt && grep -q $rand /mnt/test.txt" || volumeattachret=58
-    # cleanup so that we can run testvm without leaking volumes, IPs etc
-    nova remove-floating-ip "$instanceid" "$floatingip"
-    nova floating-ip-delete "$floatingip"
-    nova stop "$instanceid"
-    wait_for 100 1 "test \"x\$(nova show \"$instanceid\" | perl -ne 'm/ status [ |]*([a-zA-Z]+)/ && print \$1')\" == xSHUTOFF" "testvm to stop"
+    local volumeattachret=0
+
+    # do volume tests for non-docker scenario only
+    if [ -z "$want_docker" ] ; then
+        nova volume-list | grep -q available || nova volume-create 1
+        wait_for 9 5 "nova volume-list | grep available" "volume to become available" "volumecreateret=1"
+        volumeid=`nova volume-list | perl -ne "m/^[ |]*([0-9a-f-]+) [ |]*available/ && print \\$1"`
+        nova volume-attach "$instanceid" "$volumeid" /dev/vdb | tee volume-attach.out
+        volumeattachret=$?
+        device=`perl -ne "m!device [ |]*(/dev/\w+)! && print \\$1" volume-attach.out`
+        wait_for 9 5 "nova volume-show $volumeid | grep 'status.*in-use'" "volume to become attached" "volumeattachret=111"
+        ssh $vmip fdisk -l $device | grep 1073741824 || volumeattachret=$?
+        rand=$RANDOM
+        ssh $vmip "mkfs.ext3 -F $device && mount $device /mnt && echo $rand > /mnt/test.txt && umount /mnt"
+        nova volume-detach "$instanceid" "$volumeid" ; sleep 10
+        nova volume-attach "$instanceid" "$volumeid" /dev/vdb ; sleep 10
+        ssh $vmip fdisk -l $device | grep 1073741824 || volumeattachret=57
+        ssh $vmip "mount $device /mnt && grep -q $rand /mnt/test.txt" || volumeattachret=58
+        # cleanup so that we can run testvm without leaking volumes, IPs etc
+        nova remove-floating-ip "$instanceid" "$floatingip"
+        nova floating-ip-delete "$floatingip"
+        nova stop "$instanceid"
+        wait_for 100 1 "test \"x\$(nova show \"$instanceid\" | perl -ne 'm/ status [ |]*([a-zA-Z]+)/ && print \$1')\" == xSHUTOFF" "testvm to stop"
+
+        echo "Volume in VM: $volumecreateret & $volumeattachret"
+    else
+        echo "Volume in VM: tests skipped (not supported for docker)"
```

Comment:
I think I'd leave the `echo` where it was and instead do something like:
```shell
local volume_result=""

if [ -z "$want_docker" ] ; then
    local volumecreateret=0
    local volumeattachret=0
    ...
    volume_result="$volumecreateret & $volumeattachret"
else
    volume_result="tests skipped (not supported by docker backend)"
fi
...
echo "Volume in VM: $volume_result"
```

But just my personal taste, keep it the current way if you prefer the way it is.